src/Build/Instance/ProjectInstance.cs

```diff
@@ -1474,6 +1569,21 @@ public ProjectInstance DeepCopy()
             return DeepCopy(_isImmutable);
         }
 
+        /// <summary>
+        /// Create an independent clone of this object, keeping ONLY the explicitly
+        /// requested project state.
+        /// </summary>
+        /// <remarks>
+        /// Useful for reducing the wire cost of IPC for out-of-proc nodes used during
+        /// design-time builds that only need to populate a known set of data.
+        /// </remarks>
+        /// <param name="filter">Project state that should be returned.</param>
+        /// <returns></returns>
+        public ProjectInstance FilteredCopy(RequestedProjectState filter)
```

Comment:
Is there a way to only filter when we're about to send the data over the wire?  I feel that would be more straightforward where the out-of-proc node has all the data and as its writing to the named pipe its filtering out properties/items.  This seems to clone the ProjectInstance in preparation for serializing.  Do you think it would be better to do what I suggest and is it even possible with our current code?

Reply:
I don't think we can intercept only at serialization time. This approach (filter in the TargetBuilder when constructing a BuildResult) has the advantage of working in-proc as well, which is nice.

Though I guess the contract could be "return _at least_ the requested state", and returning the whole shebang in-proc would actually be faster (no deep clone).

If we were to do the filtering at the Translator level, we'd have to have the filter available in ProjectInstance, so that its TranslateAllState method could apply the filter. That's possible, but it seems like odd layering: only the result really cares about the filter; why should the ProjectInstance care?

Thoughts?